blog/posts/2020-07-09-miragejs-graphql-tools.md

```diff
@@ -0,0 +1,198 @@
+
+```js
+import { createServer } from 'miragejs';
+
+const graphqlSchema = buildSchema(`
+  type Query {
+    movies: [Movie]
+  }
+  type Movie {
+    id: ID!
+    title: String!
+  }
+`);
+
+export default function () {
+  createServer({
+    models: {
+      movie: Model
+    },
+
+    seeds(server) {
+      server.create('movie', { title: 'Interstellar' });
+      server.create('movie', { title: 'Inception' });
+      server.create('movie', { title: 'The Dark Knight' });
+    },
+
+    routes() {
+      this.post('/graphql', (schema, request) => {
+        const requestJson = JSON.parse(request.requestBody);
+        const query = requestJson.query;
+        const variables = requestJson.variables;
+
+        const resolver = {
+          movies() {
+            return schema.db.movies;
+          }
+        };
+
+        return graphql(graphqlSchema, query, resolver, null, variables);
+      });
+    }
+  })
+}
+```
+
+This is how the basic recommended configuration is when working with GraphQl. This can work early on, but become problematic to manage when you start to have a much larger schema and need to maintain way more models and relationships for your mocking. Given the following schema:
+
+```graphql
+  type Query {
+    movies: [Movie]
+  }
+  type Actor {
+    id: ID!
+    movies: [Movies]!
+    name: String!
+  }
+  type Distributor {
+    id: ID!
+    movies: [Movie]!
+    name: String!
+  }
+  type Movie {
+    id: ID!
+    actors: [Actor]!
+    distributor: Distributor!
+    title: String!
+  }
+```
+
+The first thing we can do is to automate adding models to our config at build time. This can be done by parsing our schema and some traversing the parsed AST. 
+
+```js
+import { parse } from 'graphql';
+
+const ast = parse(`
+  // ...schema
+`);
+
+// get the object definitions and fields
+const nodeTypes = ast.definitions
+  .filter(def => {
+    if (def.kind === "ObjectTypeDefinition") {
+      const { value } = def.name;
+
+      return !["Query"].includes(value);
+    }
+
+    return false;
+  })
+  .map(filteredDef => {
+    return {
+      model: filteredDef.name.value,
+      fields: filteredDef.fields
+    };
+  });
+
+// output an object with model mapping
+const modelMaps = nodeTypes.reduce((modelAccumulator, node) => {
+  modelAccumulator[node.model] = Model;
+
+  return modelAccumulator;
+}, {});
+```
+
+You can then add that do the configuration for Mirage as `models: modelMaps` and you'll get that automatically registered as you add to your schema. This does get more complicated though, when we start to add associations in our objects and need Mirage to see that as a relationship for queries that load all that data. Ideally, the graph can work for a query like so on the UI:
```

Comment:
Can we be more specific about what `that` is here? Feels weird to say add that and you'll get that.

```suggestion
We can then add that to the configuration for Mirage as `models: modelMaps` and we'll get that automatically registered as we add to our schema. This does get more complicated though, when we start to add associations in our objects and need Mirage to see that as a relationship for queries that load all that data. Ideally, the graph can work for a query like so on the UI:
```